HeterogeneousCore/SonicCore/interface/SonicClientBase.h

```diff
@@ -22,13 +25,32 @@ class SonicClientBase {
 protected:
   virtual void evaluate() = 0;
 
+  //this should be overridden by clients that allow retries
+  virtual unsigned allowedTries() const { return 0; }
+
   void setStartTime() {
+    tries_ = 0;
     if (debugName_.empty())
       return;
     t0_ = std::chrono::high_resolution_clock::now();
   }
 
-  void finish(std::exception_ptr eptr = std::exception_ptr{}) {
+  void finish(bool success, std::exception_ptr eptr = std::exception_ptr{}) {
+    if (!success) {
+      ++tries_;
+      //if max retries has not been exceeded, call evaluate again
+      if (tries_ < allowedTries()) {
+        evaluate();
```

Comment:
What should happen if the client (caller) passes an exception?

Reply:
My original thought was that allowing retries should mean allowing retries in all cases, i.e. even if an exception is passed from the client.

However, maybe it's worth adding an additional, optional parameter to denote if an exception should override remaining allowed retries, in case some exceptions are more critical than others. Alternatively, the `success` parameter could be change from a bool to an enum with allowed states "done", "try again", "stop". What do you think?

Reply:
After some further thought I find the proper behavior interesting question.

On one hand, we try to keep exceptions to signify fatal errors that should stop the job. On the other hand, it could be useful to continue retrying in some cases, but when hitting the limit throw the exception from the client. How about using `LogError`/`LogWarning` for the cases where a retry makes sense, and making the passed exception always lead to propagating it to `WaitingTaskWithArenaHolder` immediately?

Reply:
> How about using LogError/LogWarning for the cases where a retry makes sense

To make sure we're on the same page: this would just entail noting in the documentation that any exception passed to `finish()` will immediately end the process*, so if a retry is desired, client developers should emit a message instead?

* and then modifying the code to enforce this

Reply:
> > How about using LogError/LogWarning for the cases where a retry makes sense
> 
> To make sure we're on the same page: this would just entail noting in the documentation that any exception passed to `finish()` will immediately end the process*, so if a retry is desired, client developers should emit a message instead?
> 
> * and then modifying the code to enforce this

Yes (I think). Codewise here the `finish()` should first check if `eptr` is non-null, and if it is, call `holder._doneWaiting(eptr)` and return. The three states would be denoted by
* "done": `success == true`, no exception
* "try again": `success == false`, no exception (emit a message if useful)
* "stop": exception, value of `success` is irrelevant